File: apps/api/schemas/document_request.py
"""Admin-facing document request schemas."""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from apps.api.models.enums import DocumentRequestStatus, PrivateDocumentKind


class DocumentRequestItemCreate(BaseModel):
    kind: PrivateDocumentKind
    label: str
    is_required: bool = True
    ordering: int = 0


class DocumentRequestCreate(BaseModel):
    customer_id: uuid.UUID
    lead_id: uuid.UUID | None = None
    deal_id: uuid.UUID | None = None
    title: str
    instructions: str | None = None
    items: list[DocumentRequestItemCreate] = Field(min_length=1)
    expires_in_days: int = Field(default=7, ge=1, le=90)


class DocumentRequestItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    document_request_id: uuid.UUID
    kind: PrivateDocumentKind
    label: str
    is_required: bool
    ordering: int
    uploaded_document_id: uuid.UUID | None = None
    uploaded_at: datetime | None = None


class DocumentRequestResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    tenant_id: uuid.UUID
    customer_id: uuid.UUID
    lead_id: uuid.UUID | None = None
    deal_id: uuid.UUID | None = None
    title: str
    instructions: str | None = None
    token: str
    status: DocumentRequestStatus
    expires_at: datetime
    verified_at: datetime | None = None
    verification_attempts: int
    created_by_user_id: uuid.UUID
    items: list[DocumentRequestItemResponse] = []
    items_count: int = 0
    uploaded_count: int = 0
    created_at: datetime
    updated_at: datetime


class DocumentRequestCreateResponse(DocumentRequestResponse):
    """Extends base response with one-time plaintext fields."""
    verification_code: str  # plaintext — returned only on creation
    public_url: str


class DocumentRequestListResponse(BaseModel):
    items: list[DocumentRequestResponse]
    total: int


class RegenerateCodeResponse(BaseModel):
    verification_code: str
    message: str = "Verification code regenerated. Share it with the customer."

File: apps/web/src/lib/types/document-request.ts
export type DocumentRequestStatus =
  | "pending" | "partial" | "complete" | "expired" | "canceled";

export interface DocumentRequestItemResponse {
  id: string;
  document_request_id: string;
  kind: string;
  label: string;
  is_required: boolean;
  ordering: number;
  uploaded_document_id: string | null;
  uploaded_at: string | null;
}

export interface DocumentRequest {
  id: string;
  tenant_id: string;
  customer_id: string;
  lead_id: string | null;
  deal_id: string | null;
  title: string;
  instructions: string | null;
  token: string;
  status: DocumentRequestStatus;
  expires_at: string;
  verified_at: string | null;
  verification_attempts: number;
  created_by_user_id: string;
  items: DocumentRequestItemResponse[];
  items_count: number;
  uploaded_count: number;
  created_at: string;
  updated_at: string;
}

/** Returned only on create — includes plaintext code + public_url once. */
export interface DocumentRequestCreateResponse extends DocumentRequest {
  verification_code: string;
  public_url: string;
}

export interface DocumentRequestItemCreate {
  kind: string;
  label: string;
  is_required?: boolean;
  ordering?: number;
}

export interface DocumentRequestCreate {
  customer_id: string;
  title: string;
  items: DocumentRequestItemCreate[];
  lead_id?: string | null;
  deal_id?: string | null;
  property_id?: string | null;
  agent_note?: string | null;
  instructions?: string | null;
  expires_in_days?: number;
}

export interface DocumentRequestListResponse {
  items: DocumentRequest[];
  total: number;
}

export interface DocumentRequestListParams {
  status?: DocumentRequestStatus;
  customer_id?: string;
  lead_id?: string;
  deal_id?: string;
  skip?: number;
  limit?: number;
}

export interface RegenerateCodeResponse {
  verification_code: string;
  message: string;
}
